<a href="https://colab.research.google.com/github/nananke/CV/blob/gh-pages/MLCS_hw1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Name:Kenan Xu
#netID:kx2015

#step 1 Download the ling-spam dataset from http://www.aueb.gr/users/ion/data/lingspam_public.tar.gz
#Please use the “lingspam_public01” corpus with both lemmatization and stop-word enabled (under
#the lemm_stop folder).
!wget http://www.aueb.gr/users/ion/data/lingspam_public.tar.gz
!tar xvzf lingspam_public.tar.gz

Streaming output truncated to the last 5000 lines.
lingspam_public/lemm_stop/part2/6-1039msg1.txt
lingspam_public/lemm_stop/part2/6-1040msg1.txt
lingspam_public/lemm_stop/part2/6-1041msg1.txt
lingspam_public/lemm_stop/part2/6-1044msg1.txt
lingspam_public/lemm_stop/part2/6-1045msg1.txt
lingspam_public/lemm_stop/part2/6-1047msg1.txt
lingspam_public/lemm_stop/part2/6-1048msg1.txt
lingspam_public/lemm_stop/part2/6-1049msg1.txt
lingspam_public/lemm_stop/part2/6-1050msg1.txt
lingspam_public/lemm_stop/part2/6-1050msg2.txt
lingspam_public/lemm_stop/part2/6-1050msg3.txt
lingspam_public/lemm_stop/part2/6-1051msg1.txt
lingspam_public/lemm_stop/part2/6-1052msg1.txt
lingspam_public/lemm_stop/part2/6-1055msg1.txt
lingspam_public/lemm_stop/part2/6-1056msg1.txt
lingspam_public/lemm_stop/part2/spmsga142.txt
lingspam_public/lemm_stop/part2/spmsga143.txt
lingspam_public/lemm_stop/part2/spmsga144.txt
lingspam_public/lemm_stop/part2/spmsga145.txt
lingspam_public/lemm_stop/part2/spmsga146.txt
lingspam_publi

In [ ]:
#data extraction
import glob
import numpy as np
from math import log2
#create a 3-d array, d1 = part_num, d2 = file name, d3 = text content

training_dataset = []
testing_dataset = []
word_total = []
spam_count = 0
legit_count = 0
for n in range(1,10):
  subpart = "part"+str(n)
  print(subpart)
  for file in glob.glob("lingspam_public/lemm_stop/"+subpart+"/*.txt"):
    f = open(file, "r")
    d = []
    for word in f.read().split():
       #print(word)      
       d.append(word)
    word_total.append(d)

    arr = np.append([n],[file])
    f = open(file, "r")
    final_arr = np.append([arr],[f.read()])
    
   
    fullstring = file
    substring = "spm"
    if substring in fullstring:
     spam_count = spam_count + 1
     labeled_final_arr_0 = np.append([final_arr],1)
     training_dataset.append(labeled_final_arr_0)
    else:
     legit_count = legit_count + 1
     labeled_final_arr_1 = np.append([final_arr],0)
     training_dataset.append(labeled_final_arr_1)

for n in range(10,11):
  subpart = "part"+str(n)
  print(subpart)
  for file in glob.glob("lingspam_public/lemm_stop/"+subpart+"/*.txt"):
    f = open(file, "r")
    arr = np.append([n],[file])
    final_arr = np.append([arr],[f.read()])
  

     
    fullstring = file
    substring = "spm"
    if substring in fullstring:
     test_labeled_final_arr_0 = np.append([final_arr],1)
     testing_dataset.append(test_labeled_final_arr_0)
    else:
     test_labeled_final_arr_1 = np.append([final_arr],0)
     testing_dataset.append(test_labeled_final_arr_1)

print("total spam email number in training:",spam_count,"total legit email number in training", legit_count)
print("total training email number:",len(training_dataset),"total testing email number:",len(testing_dataset))





class0 = spam_count/(len(training_dataset))
class1 = legit_count/(len(training_dataset))



part1
part2
part3
part4
part5
part6
part7
part8
part9
part10
total spam email number in training: 432 total legit email number in training 2170
total training email number: 2602 total testing email number: 291


In [ ]:
#information gain calculation

unique_total_word = []
for x in range(0,len(word_total)):
  for y in range(0,len(word_total[x])):
    unique_total_word.append(word_total[x][y])

#delete the duplicate one for the total word array in order to decease the running time
unique_total_word = list(dict.fromkeys(unique_total_word))
unique_total_word.remove("!")
#array to save info gain result
gain_result = []
#entropy before the given word 
hc = -(class0 * log2(class0) + class1 * log2(class1))
#print(hc)

for y in range (0, len(unique_total_word)):
  print("processing: ",y,"/",len(unique_total_word))
  #legit word appear
  lwa =0  
  #legit word not appear
  lwna =0
  #spam word appear
  swa =0
  #spam word not appear
  swna =0

  for x in range(0,len(training_dataset)):
   #split sentence into bag of words first, then do the comparsion
   words = training_dataset[x][2].split()
   if unique_total_word[y] in words:
    if training_dataset[x][3] == '1':
     swa = swa + 1
     #print('found in spam: ',training_dataset[x][3])
    else :
      lwa = lwa + 1  
      #print('found in legit',training_dataset[x][3])

  lwna = 2170 - lwa
  swna = 432 - swa
  #print(lwa,lwna,swa,swna)
  
  try:
    entropy1 = -((lwa)/(lwa+swa) * log2((lwa)/(lwa+swa)) + (swa)/(lwa+swa) * log2((swa)/(lwa+swa)))
    entropy2 = -((lwna)/(lwna+swna) * log2((lwna)/(lwna+swna)) + (swna)/(lwna+swna) * log2((swna)/(lwna+swna)))
  except:
    #since there are some word with P = 0, using Laplace smoothing to ensure there will be no math error
    entropy1 = -((lwa+1)/(lwa+swa+2) * log2((lwa+1)/(lwa+swa+2)) + (swa+1)/(lwa+swa+2) * log2((swa+1)/(lwa+swa+2)))
    entropy2 = -((lwna+1)/(lwna+swna+2) * log2((lwna+1)/(lwna+swna+2)) + (swna+1)/(lwna+swna+2) * log2((swna+1)/(lwna+swna+2)))

  #H(c|x)
  hcx =  (lwa+swa)/len(training_dataset)*entropy1 + (lwna+swna)/len(training_dataset)*entropy2 

  #IG = h(c) - h(c|x)
  info_gain = hc - hcx
  m = np.append([info_gain],[unique_total_word[y]])
  gain_result.append(m)
  #print(gain_result," ", unique_total_word[y])

Streaming output truncated to the last 5000 lines.
processing:  50769 / 55768
processing:  50770 / 55768
processing:  50771 / 55768
processing:  50772 / 55768
processing:  50773 / 55768
processing:  50774 / 55768
processing:  50775 / 55768
processing:  50776 / 55768
processing:  50777 / 55768
processing:  50778 / 55768
processing:  50779 / 55768
processing:  50780 / 55768
processing:  50781 / 55768
processing:  50782 / 55768
processing:  50783 / 55768
processing:  50784 / 55768
processing:  50785 / 55768
processing:  50786 / 55768
processing:  50787 / 55768
processing:  50788 / 55768
processing:  50789 / 55768
processing:  50790 / 55768
processing:  50791 / 55768
processing:  50792 / 55768
processing:  50793 / 55768
processing:  50794 / 55768
processing:  50795 / 55768
processing:  50796 / 55768
processing:  50797 / 55768
processing:  50798 / 55768
processing:  50799 / 55768
processing:  50800 / 55768
processing:  50801 / 55768
processing:  50802 / 55768
processing:  50803 / 55768
proc

In [ ]:
#saving gain_result to local
'''
with open('your_file.txt', 'w') as f:
    for item in gain_result:
        f.write("%s\n" % item)
'''

In [ ]:
#once we have calculate the IG, we can just upload txt from our local file since it takes long time to re-calculate
'''
gain_result=[]
import re
f = open("your_file.txt", "r")
for x in f:
  extract = re.findall(r"'(.*?)'", x, re.DOTALL)
  try:
   #print(extract[1])
   gain_result.append(extract)
  except:
    None
'''

In [ ]:

print('TOP 10 Feature words List')

for x in range(0,len(gain_result)):
  gain_result[x][0] = '{:.5f}'.format(float(gain_result[x][0]))
gain_result.sort(reverse=False,key=lambda x: x[0]) 
for x in range(len(gain_result)-10,len(gain_result)):
  print(gain_result[x])

TOP 10 Feature words List
['0.08879' 'our']
['0.09251' 'market']
['0.09418' '@']
['0.10117' 'click']
['0.11868' 'money']
['0.14501' 'university']
['0.14515' 'linguistic']
['0.16573' 'free']
['0.16891' 'remove']
['0.20444' 'language']


In [ ]:
from sklearn.naive_bayes import BernoulliNB
#first top 10,100,1000 features with BernoulliNB
feature_10 = []
feature_100 = []
feature_1000 = []
for x in range(len(gain_result)-10,len(gain_result)):
  #print(gain_result[x])
  feature_10.append(gain_result[x][1])

for x in range(len(gain_result)-100,len(gain_result)):
  #print(gain_result[x])
  feature_100.append(gain_result[x][1])

for x in range(len(gain_result)-1000,len(gain_result)):
  #print(gain_result[x])
  feature_1000.append(gain_result[x][1])

feature_array = [feature_10,feature_100,feature_1000]


In [ ]:

#Bernoulli NB classifier with binary features
#for each file in training text, feature is found =1, feature not found =0
for m in range (0, 3):
 X_train =[]
 y_train =[]
 X_test = []
 y_test = []
 for x in range(0,len(training_dataset)): 
   row = [0] * len(feature_array[m])
   y_train.append(int(training_dataset[x][3]))
   wordsbag = training_dataset[x][2].split()
   for y in range (0,len(feature_array[m])):    
     if feature_array[m][y] in wordsbag:
       row[y] = 1 
   X_train.append(row)
 
 #for each file in training text, feature is found =1, feature not found =0
 for x in range(0,len(testing_dataset)): 
   row = [0] * len(feature_array[m])
   y_test.append(int(testing_dataset[x][3]))
   for y in range (0,len(feature_array[m])):    
     if feature_array[m][y] in testing_dataset[x][2]:
       row[y] = 1 
   X_test.append(row)
   

 from sklearn.naive_bayes import BernoulliNB
 from sklearn.model_selection import train_test_split
 from sklearn import metrics
 from sklearn.metrics import confusion_matrix
 # Create three binary features
 X = X_train
 # Create a binary target vector
 y = y_train
 # View first ten observations
 
 # Create Bernoulli Naive Bayes object with prior probabilities of each class
 clf = BernoulliNB(class_prior=[0.25, 0.5])
 
 # Train model
 model = clf.fit(X, y)
 #print(len(clf.predict(X_test)))
 #print(len(y_test))
 tn, fp, fn, tp = confusion_matrix(y_test,model.predict(X_test)).ravel()
 print('NB classifier with binary features',len(feature_array[m]),'features ','Spam precision:',tn/(tn+fn))
 print('NB classifier with binary features',len(feature_array[m]),'features ','Spam recall',clf.score(X_test, y_test))


NB classifier with binary features 10 features  Spam precision: 0.9869565217391304
NB classifier with binary features 10 features  Spam recall 0.9381443298969072
NB classifier with binary features 100 features  Spam precision: 0.9714285714285714
NB classifier with binary features 100 features  Spam recall 0.9621993127147767
NB classifier with binary features 1000 features  Spam precision: 0.9444444444444444
NB classifier with binary features 1000 features  Spam recall 0.9381443298969072


In [ ]:
#Multinomial NB with binary features;
#Multi NB shares the feature training data with Ber NB since they all needs binary data 
#for each file in training text, feature is found =1, feature not found =0
#for each file in training text, feature is found =1, feature not found =0
for m in range (0, 3):
 X_train =[]
 y_train =[]
 X_test = []
 y_test = []
 for x in range(0,len(training_dataset)): 
   row = [0] * len(feature_array[m])
   y_train.append(int(training_dataset[x][3]))
   wordsbag = training_dataset[x][2].split()
   for y in range (0,len(feature_array[m])):    
     if feature_array[m][y] in wordsbag:
       row[y] = 1 
   X_train.append(row)
 
 #for each file in training text, feature is found =1, feature not found =0
 for x in range(0,len(testing_dataset)): 
   row = [0] * len(feature_array[m])
   y_test.append(int(testing_dataset[x][3]))
   for y in range (0,len(feature_array[m])):    
     if feature_array[m][y] in testing_dataset[x][2]:
       row[y] = 1 
   X_test.append(row)
   

 
 from sklearn.naive_bayes import MultinomialNB 
 # Create three binary features
 X = X_train
 # Create a binary target vector
 y = y_train
 # View first ten observations
 
 # Create Bernoulli Naive Bayes object with prior probabilities of each class
 clf = MultinomialNB()
 
 # Train model
 model = clf.fit(X, y)
 #print(len(clf.predict(X_test)))
 #print(len(y_test))
 tn, fp, fn, tp = confusion_matrix(y_test,model.predict(X_test)).ravel()
 print('Multinomial NB with binary features',len(feature_array[m]),'features ','Spam precision:',tn/(tn+fn))
 print('Multinomial NB with binary features',len(feature_array[m]),'features ','Spam recall',clf.score(X_test, y_test))

Multinomial NB with binary features 10 features  Spam precision: 0.9598393574297188
Multinomial NB with binary features 10 features  Spam recall 0.9553264604810997
Multinomial NB with binary features 100 features  Spam precision: 0.9836734693877551
Multinomial NB with binary features 100 features  Spam recall 0.9828178694158075
Multinomial NB with binary features 1000 features  Spam precision: 0.9758064516129032
Multinomial NB with binary features 1000 features  Spam recall 0.979381443298969


In [ ]:
#Multi TF, we need to re-caculate training data to count for how many times feature words appear
for m in range (0, 3):
 X_train =[]
 y_train =[]
 X_test = []
 y_test = []
 for x in range(0,len(training_dataset)): 
   row = [0] * len(feature_array[m])
   y_train.append(int(training_dataset[x][3]))
   wordsbag = training_dataset[x][2].split()
   for y in range (0,len(feature_array[m])):          
       count = wordsbag.count(feature_array[m][y])
       row[y] = count 
   X_train.append(row)


  
 
 #for each file in training text, feature is found =1, feature not found =0
 for x in range(0,len(testing_dataset)): 
   row = [0] * len(feature_array[m])
   y_test.append(int(testing_dataset[x][3]))
   for y in range (0,len(feature_array[m])):    
       count = testing_dataset[x][2].count(feature_array[m][y])
       row[y] = count 
   X_test.append(row)
 

 from sklearn.naive_bayes import MultinomialNB 

 # Create three binary features
 X = X_train
 # Create a binary target vector
 y = y_train
 # View first ten observations
 
 # Create Bernoulli Naive Bayes object with prior probabilities of each class
 clf = MultinomialNB()
 # Train model
 model = clf.fit(X, y)
 #print(len(clf.predict(X_test)))
 #print(len(y_test))

 tn, fp, fn, tp = confusion_matrix(y_test,model.predict(X_test)).ravel()
 print('Multi TF:',len(feature_array[m]),'features ','Spam precision:',tn/(tn+fn))
 print('Multi TF:',len(feature_array[m]),'features ','Spam recall',clf.score(X_test, y_test))

Multi TF: 10 features  Spam precision: 0.9871794871794872
Multi TF: 10 features  Spam recall 0.9518900343642611
Multi TF: 100 features  Spam precision: 0.9795918367346939
Multi TF: 100 features  Spam recall 0.9759450171821306
Multi TF: 1000 features  Spam precision: 0.8515901060070671
Multi TF: 1000 features  Spam recall 0.852233676975945


In [ ]:
#SVM based spam filter
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
for m in range (0, 3):
 X_train =[]
 y_train =[]
 for x in range(0,len(training_dataset)): 
   row = [0] * len(feature_array[m])
   y_train.append(int(training_dataset[x][3]))
   wordsbag = training_dataset[x][2].split()
   for y in range (0,len(feature_array[m])):          
       count = wordsbag.count(feature_array[m][y])
       row[y] = count 
   X_train.append(row)

 X_test = []
 y_test = []
  
 
 #for each file in training text, feature is found =1, feature not found =0
 for x in range(0,len(testing_dataset)): 
   row = [0] * len(feature_array[m])
   y_test.append(int(testing_dataset[x][3]))
   for y in range (0,len(feature_array[m])):    
       count = testing_dataset[x][2].count(feature_array[m][y])
       row[y] = count 
   X_test.append(row)

 X = X_train
 y = y_train
 from sklearn.svm import SVC
 from sklearn.model_selection import GridSearchCV
 
 #cross-validation on the training dataset.
 gamma = []
 i = 0.0001
 while i > 0.000000001:
  gamma.append(i) 
  i = i / 10
 param_grid = [
  {'C': [0.1, 1000, 10000,100000], 'kernel': ['linear']},
  {'C': [0.1, 1000, 10000,100000], 'gamma': gamma, 'kernel': ['rbf']},
  ]
 clf = GridSearchCV(SVC(), param_grid, cv=3, refit=True, verbose=100,return_train_score = True)
 clf.fit(X, y)

 from sklearn.metrics import accuracy_score
 print(clf.best_params_)
 y_pred = clf.predict(X_test)
 accuracy_score(y_pred, y_test)
 tn, fp, fn, tp = confusion_matrix(y_test,clf.predict(X_test)).ravel()
 print(len(feature_array[m]),'features ','Spam precision:',tn/(tn+fn))
 print(len(feature_array[m]),'features ','Spam recall',clf.score(X_test, y_test))
 #feature 10
 #{'C': 100000, 'gamma': 0.0001, 'kernel': 'rbf'}
 #10 features  Spam precision: 0.8222222222222222
 #10 features  Spam recall 0.9312714776632303

 #feature100
 #{'C': 0.1, 'kernel': 'linear'}
 #100 features  Spam precision: 0.7
 #100 features  Spam recall 0.9140893470790378

 #feature1000
 #{'C': 100000, 'gamma': 1.0000000000000002e-07, 'kernel': 'rbf'}
 #1000 features  Spam precision: 1.0
 #1000 features  Spam recall 0.8625429553264605

Fitting 3 folds for each of 28 candidates, totalling 84 fits
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[CV] C=0.1, kernel=linear ............................................
[CV]  C=0.1, kernel=linear, score=(train=0.962, test=0.960), total=   0.0s
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[CV] C=0.1, kernel=linear ............................................
[CV]  C=0.1, kernel=linear, score=(train=0.957, test=0.957), total=   0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.0s remaining:    0.0s
[CV] C=0.1, kernel=linear ............................................
[CV]  C=0.1, kernel=linear, score=(train=0.963, test=0.947), total=   0.0s
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    0.1s remaining:    0.0s
[CV] C=1000, kernel=linear ...........................................
[CV]  C=1000, kernel=linear, score=(train=0.971, test=0.956), total=   1.0s
[Parallel(n_jobs=1)]: Done   4 ou

In [ ]:
#Adversarial Classification
#Bernoulli NB classifier with binary features
#for each file in training text, feature is found =1, feature not found =0
'''
for m in range (0, 1):
 X_train =[]
 y_train =[]
 X_test = []
 y_test = []
 for x in range(0,len(training_dataset)): 
   row = [0] * len(feature_array[m])
   y_train.append(int(training_dataset[x][3]))
   wordsbag = training_dataset[x][2].split()
   for y in range (0,len(feature_array[m])):    
     if feature_array[m][y] in wordsbag:
       row[y] = 1 
   X_train.append(row)
 
 #for each file in training text, feature is found =1, feature not found =0
 for x in range(0,len(testing_dataset)): 
   row = [0] * len(feature_array[m])
   y_test.append(int(testing_dataset[x][3]))
   for y in range (0,len(feature_array[m])):    
     if feature_array[m][y] in testing_dataset[x][2]:
       row[y] = 1 
   X_test.append(row)
   

 from sklearn.naive_bayes import BernoulliNB
 from sklearn.model_selection import train_test_split
 from sklearn import metrics
 from sklearn.metrics import confusion_matrix
 # Create three binary features
 X = X_train
 # Create a binary target vector
 y = y_train
 # View first ten observations
 
 # Create Bernoulli Naive Bayes object with prior probabilities of each class
 clf = BernoulliNB(class_prior=[0.25, 0.5])
 
 # Train model
 model = clf.fit(X, y)
 #print(len(clf.predict(X_test)))
 #print(len(y_test))
 tn, fp, fn, tp = confusion_matrix(y_test,model.predict(X_test)).ravel()
 #false negative rate
 print(len(feature_array[m]),'features ','false negative rate before ad. modify:',fn/(fn+tp))
'''


10 features  false negative rate before ad. modify: 0.01652892561983471


In [ ]:
#spam number given x
#calculation W(gapx) for each feature word
gapx_array = []
nenenenen = []
for y in range (len(gain_result)-10,len(gain_result)):
 x_length = 0
 spam_num_x = 0
 legit_num_x = 0
 for x in range(0,len(training_dataset)):
  
   #split sentence into bag of words first, then do the comparsion
   words = training_dataset[x][2].split()
   if gain_result[y][1] in words:
    x_length = x_length + 1 
    if training_dataset[x][3] == '1':
     spam_num_x = spam_num_x + 1
     #print('found in spam: ',training_dataset[x][3])
    else :
      legit_num_x = legit_num_x + 1  
      #print('found in legit',training_dataset[x][3])
 print(x_length,spam_num_x,legit_num_x)

 nenenenen = gain_result[y]
 nenenenen.append((spam_num_x/legit_num_x)-1)
 nenenenen = list(dict.fromkeys(nenenenen))
 gapx_array.append(nenenenen)


627 260 367
173 132 41
1671 107 1564
129 116 13
227 169 58
1293 17 1276
1105 0 1105
380 251 129
207 185 22
1454 7 1447


In [ ]:

#P(xi|+) = how many spam email contains word feature
#P(xi|-) = how many legit email contains word feature

for y in range (0,len(gapx_array)):
 legit_num_x = 0
 spam_num_x = 0
 for x in range(0,len(training_dataset)):  
   #split sentence into bag of words first, then do the comparsion
   words = training_dataset[x][2].split()
   if gapx_array[y][1] in words:
    if training_dataset[x][3] == '1':
     spam_num_x = spam_num_x + 1
     #print('found in spam: ',training_dataset[x][3])
    else :
      legit_num_x = legit_num_x + 1  
      #print('found in legit',training_dataset[x][3])
 print(spam_num_x,legit_num_x)


 #we define lcox -> locxi as 0 -> 1 (good word appear) to overcome the 'bad' word influence
 try:
  locx = log2((432 - spam_num_x)/(2170-legit_num_x)/432*2170)
 except:
  locx = log2((432 - spam_num_x+1)/(2170-legit_num_x+2)/432*2170)
 try:
  locxi= log2(spam_num_x/legit_num_x/432*2170)
 except:
  locxi= log2((spam_num_x+1)/(legit_num_x+2)/432*2170)

 delta_loc = locx - locxi 


 gapx_array[y].append(delta_loc)
#total spam email number in training: 432 total legit email number in training 2170
#total training email number: 2602 total testing email number: 291



260 367
132 41
107 1564
116 13
169 58
17 1276
0 1105
251 129
185 22
7 1447


In [ ]:
#explain each element in gapx_ array:
#0 = IG value
#1 = word feature\
#2 = gap(x) value P(+|x)/P(-|x) - 1 
#3 = delta Loc = Loc(x) - Loc(x') ， we only consider postive value here, since negative value meaning switching the term does not help 
for x in range (0,len(gapx_array)):
  print(gapx_array[x])



['0.08879', 'our', -0.29155313351498635, -2.8926504869154854]
['0.09251', 'market', 2.2195121951219514, -4.513983658696782]
['0.09418', '@', -0.9315856777493606, 2.970679735349072]
['0.10117', 'click', 7.923076923076923, -5.928570989969831]
['0.11868', 'money', 1.9137931034482758, -4.548373571220764]
['0.14501', 'university', -0.9866771159874608, 5.122786277565499]
['0.14515', 'linguistic', -1.0, 8.810689293831821]
['0.16573', 'free', 0.945736434108527, -4.455530878477032]
['0.16891', 'remove', 7.409090909090908, -6.192360888279749]
['0.20444', 'language', -0.995162404975812, 6.924961478548067]


In [ ]:

##Adversarial STRATEGY
#MCC(x) is the nearest instance (costwise) to x which naive Bayes classifies as negative

#Since top-10 features all has positive IG value, we do not have to check if delta Loc >=0


'''
x = 8
w = gapx_array[x][2]    
find_MCC(x,w)
print(MinCost,MinList)

def find_MCC(x,w):
  # loop through n = (0,9)
 if w <= 0:
  return 0,[]
 if x == 0:
   return 100000, None

 MinCost = 100000 
 MinList =  []

 #for x in range (0,7):
 print(gapx_array[x])
 if gapx_array[x][3] >=0:
    CurCost, CurList = find_MCC(x-1,float(w)-float(gapx_array[x][3]))
    CurCost = CurCost + 1
    List_adding = []
    List_adding.append(gapx_array[x][1])
    List_adding.append(x)
    CurList = List_adding

    if CurCost < MinCost:     
      MinCost = CurCost      
      MinList = CurList
      #print(MinCost,MinList)
 return MinCost, MinList
 
'''
#more efficient way to find the optimal solution for each feature word that gapx > 0 
filter_postive = []
for x in range (0,len(gapx_array)):
 if gapx_array[x][2] > 0:
   filter_postive.append(gapx_array[x])


filter_negative = []
for x in range (0,len(gapx_array)):
 if gapx_array[x][3] >= 0:
   #print(gapx_array[x])
   filter_negative.append(gapx_array[x])



filter_negative.sort(reverse=True,key=lambda x: x[3])


total_gap = []
total_word = []
import itertools
for L in range(0, len(filter_postive)+1):
  #get all the combinations of words that gap(x) is larger than 0 
    for subset in itertools.combinations(filter_postive, L):
      if len(subset) > 0:
       # print(subset)
        count = 0
        word_count = []
        for z in range(0,len(subset)):
          count = count + subset[z][2]
          word_count.append(subset[z][1])
        total_gap.append(count)
        total_word.append(word_count)
filter_postive.sort(reverse=True,key=lambda x: x[2])

final_gap_word_count = []
for x in range (0,len(total_gap)):
  temp = []
  temp.append(total_word[x])
  
  n = 0
  mins_count = filter_negative[n][3]
  while total_gap[x] - mins_count > 0:
    n = n +1
    mins_count= mins_count + filter_negative[n][3]
  temp.append(n+1)
  #final_gap_count store all the combination of 'spam' word and how many 'legit' word (sorted in a descending order) need to add to make w < 0
  final_gap_word_count.append(temp)

In [ ]:
for x in range (0,len(final_gap_word_count)):
  print(final_gap_word_count[x])

[['market'], 1]
[['click'], 1]
[['money'], 1]
[['free'], 1]
[['remove'], 1]
[['market', 'click'], 2]
[['market', 'money'], 1]
[['market', 'free'], 1]
[['market', 'remove'], 2]
[['click', 'money'], 2]
[['click', 'free'], 2]
[['click', 'remove'], 2]
[['money', 'free'], 1]
[['money', 'remove'], 2]
[['free', 'remove'], 1]
[['market', 'click', 'money'], 2]
[['market', 'click', 'free'], 2]
[['market', 'click', 'remove'], 3]
[['market', 'money', 'free'], 1]
[['market', 'money', 'remove'], 2]
[['market', 'free', 'remove'], 2]
[['click', 'money', 'free'], 2]
[['click', 'money', 'remove'], 3]
[['click', 'free', 'remove'], 3]
[['money', 'free', 'remove'], 2]
[['market', 'click', 'money', 'free'], 2]
[['market', 'click', 'money', 'remove'], 3]
[['market', 'click', 'free', 'remove'], 3]
[['market', 'money', 'free', 'remove'], 2]
[['click', 'money', 'free', 'remove'], 3]
[['market', 'click', 'money', 'free', 'remove'], 3]


From above result, it shows any combinations of 'bad' words in 10 features, and the number(also represent for the cost) represents how many 'good' words has to be add, accroding to good words in a descending order, to make this email being legit to classifier

#I replace the recursive function with another approach that will reduce the running time, here is my explaination:
# firstly, I made 2 arrays from 10 features. One array, which called filter_postive store the features that gap(x),w > 0.
#Second array called filter_negative that store features that has positive ∆LOi value. Then sorting the second array in descending order, thus, we can do what inside the recursive function:(w − ∆LOi) and get the minimum result list when w < 0 in the fastest way.

In [ ]:
#Add word to testing dataset to see if it works

for m in range (0, 1):
 X_train =[]
 y_train =[]
 X_test = []
 y_test = []
 for x in range(0,len(training_dataset)): 
   row = [0] * len(feature_array[m])
   y_train.append(int(training_dataset[x][3]))
   wordsbag = training_dataset[x][2].split()
   for y in range (0,len(feature_array[m])):    
     if feature_array[m][y] in wordsbag:
       row[y] = 1 
   X_train.append(row)
 
 #for each file in training text, feature is found =1, feature not found =0
 for x in range(0,len(testing_dataset)): 
   row = [0] * len(feature_array[m])
   y_test.append(int(testing_dataset[x][3]))
   wordsbag = testing_dataset[x][2].split()
   for y in range (0,len(feature_array[m])):    
     if feature_array[m][y] in wordsbag:
       row[y] = 1 

   
   X_test.append(row)
   

 from sklearn.naive_bayes import BernoulliNB
 from sklearn.model_selection import train_test_split
 from sklearn import metrics
 from sklearn.metrics import confusion_matrix
 # Create three binary features
 X = X_train
 # Create a binary target vector
 y = y_train
 # View first ten observations
 
 # Create Bernoulli Naive Bayes object with prior probabilities of each class
 clf = BernoulliNB(class_prior=[0.25, 0.5])
 
 # Train model
 model = clf.fit(X, y)
 #print(len(clf.predict(X_test)))
 #print(len(y_test))
 tn, fp, fn, tp = confusion_matrix(y_test,model.predict(X_test)).ravel()
 #false negative rate
 print(len(feature_array[m]),'features ','false negative rate BEFORE ad. modify:',fn/(fn+tp))
 print(len(feature_array[m]),'features ','Spam recall',clf.score(X_test, y_test))
 #print(len(feature_array[m]),'features ','false positive rate Before ad. modify:',fp/(fp+tn))


10 features  false negative rate BEFORE ad. modify: 0.04081632653061224
10 features  Spam recall 0.9518900343642611


In [ ]:
new_X_test = X_test
y_pred = clf.predict(X_test)
#print(len(y_pred),len(new_X_test))
last_cost_result = []
final_gap_word_count.sort(reverse=True,key=lambda x: len(x[0]))

for x in range (0,len(testing_dataset)):
 if y_pred[x] == 1 :
  wordsbag  =  testing_dataset[x][2].split()
  for y in range (0,(len(final_gap_word_count))):
    if set(final_gap_word_count[y][0]).issubset(set(wordsbag)) == True:
      newsdsfdsf = []
      newsdsfdsf.append(final_gap_word_count[y])
      newsdsfdsf.append(x)
      last_cost_result.append(newsdsfdsf)
      break
  
last_cost_result.sort(reverse=True,key=lambda x: x[1])
an_iterator = itertools.groupby(last_cost_result, lambda x : x[1]) 




p_count = 0
#algo. 2
#since we already knew the feature word 'linguistic' has largest delta Loc value
# and this feature word can overcome any feature words that w > 0 with cost 1 
# therefore, following the algo. #2, we add this 'linguistic' to any email if NB(x) = +

'''
for x in range (0,len(new_X_test)):
  if y_pred[x] == 1:
    new_X_test[x][6] = 1
    new_X_test[x][9] = 1
    new_X_test[x][5] = 1
'''

#index 6 for each row in X_test is the position for word 'linguistic' (largest impact)
#index 9 for each row in X_test is the position for word 'language' (second largest impact)
#index 5 for each row in X_test is the position for word 'university'(third largest impact)

cost = 0
total_spam_count = 0
for x in range (0,len(last_cost_result)):
  if last_cost_result[x][0][1] == 1:
    new_X_test[last_cost_result[x][1]][6] = 1   
    cost = cost + 1 
    total_spam_count = total_spam_count + 1 
  elif last_cost_result[x][0][1] == 2:
    new_X_test[last_cost_result[x][1]][6] = 1
    new_X_test[last_cost_result[x][1]][9] = 1
    cost = cost + 2 
    total_spam_count = total_spam_count + 1 
  elif last_cost_result[x][0][1] == 3:
    new_X_test[last_cost_result[x][1]][6] = 1
    new_X_test[last_cost_result[x][1]][9] = 1
    new_X_test[last_cost_result[x][1]][5] = 1
    cost = cost + 3
    total_spam_count = total_spam_count + 1 


second_part = []
for x in range(0,len(last_cost_result)):
  second_part.append(last_cost_result[x][1])


#somehow ther is a few email still classified as spam even after adding the 'largest impact word', so I treat those email individually.
# and final accuray increased after adding 3 'impact words'
new_y_pred = clf.predict(new_X_test)
for x in range (0,len(new_y_pred)):
  if new_y_pred[x] == 1 :
    new_X_test[x][6] = 1
    new_X_test[x][9] = 1
    new_X_test[x][5] = 1
    cost = cost + 3
    for y in range (0,len(last_cost_result)):
      if x == last_cost_result[y][1]:
        #when we treating those speical email (those still classified as spam even after adding the 'largest impact word'), we do not want the cost be calculated twice, therefore, we have to minus the orginal cost that we added in the previous part
        cost = cost - last_cost_result[y][0][1]
       
    total_spam_count = total_spam_count + 1



     


print('average cost over all email: ',cost/total_spam_count)

print('accuracy recall after adversary strategy:', clf.score(new_X_test, y_test))
tn, fp, fn, tp = confusion_matrix(y_test,clf.predict(new_X_test)).ravel()
#print(fn, tp, tn, fp)
print(len(feature_array[m]),'features ','false negative rate AFTER ad. modify:',fn/(fn+tp))
#print(len(feature_array[m]),'features ','false positive rate AFTER ad. modify:',tp/(tp+fn))
#from here, we found the false positive rate is 0, which means all the spam was falsely predicated as positive



average cost over all email:  1.7272727272727273
accuracy recall after adversary strategy: 0.8419243986254296
10 features  false negative rate AFTER ad. modify: 0.9387755102040817


In [ ]:
#CLASSIFIER STRATEGY
#for all negative prediction, using a greedy approach, let's minus its total_gap(x) from the largest gap(x) value word, and store how many words were removed
y_new_pred = clf.predict(new_X_test)
final_y_pred = y_new_pred
gap_value = []
y_new_pred = clf.predict(new_X_test)
for y in range (0,len(new_X_test)):
  gap_value_count = 0
  if y_new_pred[y] == 0:  
    for x in range (0,len(new_X_test[y])):
      if new_X_test[y][x] == 1:
        gap_value_count = gap_value_count + gapx_array[x][2]
  gap_value.append(gap_value_count)
  #print(gap_value_count,y_new_pred[y],y_test[y])


for x in range (0,len(gap_value)):
  #bad_key_word_count = 0 
 # good_key_word_count= 0
  if y_new_pred[x] == 0:
    if new_X_test[x][6] == 1 :
    #  good_key_word_count = good_key_word_count + 1
      if gap_value[x] >= 0 :
        final_y_pred[x] = 1
  #print(gap_value[x],final_y_pred[x],y_test[x])

for x in range(0,len(new_X_test)):
  bad_key_word_count = 0 
  good_key_word_count= 0

  if new_X_test[x][6] == 1 :
      good_key_word_count = good_key_word_count + 1 
  if new_X_test[x][9] == 1 :
      good_key_word_count = good_key_word_count + 1 
  if new_X_test[x][1] == 1:
      bad_key_word_count = bad_key_word_count + 1
  if new_X_test[x][3] == 1:
      bad_key_word_count = bad_key_word_count + 1
  if new_X_test[x][8] == 1:
      bad_key_word_count = bad_key_word_count + 1
  if new_X_test[x][4] == 1:
      bad_key_word_count = bad_key_word_count + 1
  if bad_key_word_count >= 2 :
    final_y_pred[x] ==1
    
from sklearn.metrics import accuracy_score
print('accuracy after classifier response',accuracy_score(y_test, final_y_pred))
tn, fp, fn, tp = confusion_matrix(y_test,final_y_pred).ravel()
print(len(feature_array[m]),'features ','false negative rate AFTER ad. modify:',fn/(fn+tp))
  

accuracy after classifier response 0.9037800687285223
10 features  false negative rate AFTER ad. modify: 0.3673469387755102


In [ ]:
'''
for x in range(0,len(new_X_test)):
  bad_key_word_count = 0 
  bad_key_word_value = 0
  good_key_word_count= 0
  good_key_word_value = 0

  if new_X_test[x][6] == 1 :
      good_key_word_count = good_key_word_count + 1 
      good_key_word_value = good_key_word_value + gapx_array[6][3]
  if new_X_test[x][9] == 1 :
      good_key_word_count = good_key_word_count + 1 
      good_key_word_value = good_key_word_value + gapx_array[9][3]
  if new_X_test[x][1] == 1:
      bad_key_word_count = bad_key_word_count + 1
      bad_key_word_value = bad_key_word_value + gapx_array[1][2]
  if new_X_test[x][3] == 1:
      bad_key_word_count = bad_key_word_count + 1
      bad_key_word_value = bad_key_word_value + gapx_array[3][2]
  if new_X_test[x][8] == 1:
      bad_key_word_count = bad_key_word_count + 1
      bad_key_word_value = bad_key_word_value + gapx_array[8][2]
  if new_X_test[x][4] == 1:
      bad_key_word_count = bad_key_word_count + 1
      bad_key_word_value = bad_key_word_value + gapx_array[4][2]
  print('bad:',bad_key_word_count,'good:',good_key_word_count,'pred:',final_y_pred[x],'acutal:',y_test[x],bad_key_word_value,good_key_word_value)   
'''

"\nfor x in range(0,len(new_X_test)):\n  bad_key_word_count = 0 \n  bad_key_word_value = 0\n  good_key_word_count= 0\n  good_key_word_value = 0\n\n  if new_X_test[x][6] == 1 :\n      good_key_word_count = good_key_word_count + 1 \n      good_key_word_value = good_key_word_value + gapx_array[6][3]\n  if new_X_test[x][9] == 1 :\n      good_key_word_count = good_key_word_count + 1 \n      good_key_word_value = good_key_word_value + gapx_array[9][3]\n  if new_X_test[x][1] == 1:\n      bad_key_word_count = bad_key_word_count + 1\n      bad_key_word_value = bad_key_word_value + gapx_array[1][2]\n  if new_X_test[x][3] == 1:\n      bad_key_word_count = bad_key_word_count + 1\n      bad_key_word_value = bad_key_word_value + gapx_array[3][2]\n  if new_X_test[x][8] == 1:\n      bad_key_word_count = bad_key_word_count + 1\n      bad_key_word_value = bad_key_word_value + gapx_array[8][2]\n  if new_X_test[x][4] == 1:\n      bad_key_word_count = bad_key_word_count + 1\n      bad_key_word_value = bad_

Final Report:

1.Top-10 words from IG result:

  ['0.08879', 'our']<br>
  ['0.09251', 'market']<br>
  ['0.09418', '@']<br>
  ['0.10117', 'click']<br>
  ['0.11868', 'money']<br>
  ['0.14501', 'university']<br>
  ['0.14515', 'linguistic']<br>
  ['0.16573', 'free']<br>
  ['0.16891', 'remove']<br>
  ['0.20444', 'language']<br>


<br>2. 9 rows:<br>
NB classifier with binary features 10 features  Spam precision: 0.9869565217391304<br>
NB classifier with binary features 10 features  Spam recall 0.9381443298969072<br>
NB classifier with binary features 100 features  Spam precision: 0.9714285714285714<br>
NB classifier with binary features 100 features  Spam recall 0.9621993127147767<br>
NB classifier with binary features 1000 features  Spam precision: 0.944<br>
NB classifier with binary features 1000 features  Spam recall 0.9381443298969072<br>
<br>
Multinomial NB with binary features 10 features  Spam precision: 0.9598393574297188<br>
Multinomial NB with binary features 10 features  Spam recall 0.9553264604810997<br>
Multinomial NB with binary features 100 features  Spam precision: 0.9836734693877551<br>
Multinomial NB with binary features 100 features  Spam recall 0.9828178694158075<br>
Multinomial NB with binary features 1000 features  Spam precision: 0.9758064516129032<br>
Multinomial NB with binary features 1000 features  Spam recall 0.979381443298969<br>
<br>
Multi TF: 10 features  Spam precision: 0.9871794871794872<br>
Multi TF: 10 features  Spam recall 0.9518900343642611<br>
Multi TF: 100 features  Spam precision: 0.9795918367346939<br>
Multi TF: 100 features  Spam recall 0.9759450171821306<br>
Multi TF: 1000 features  Spam precision: 0.8515901060070671<br>
Multi TF: 1000 features  Spam recall 0.852233676975945<br>

3.SVM<br>
From SVM, I tested top10,top100,top1000 features<br>
I used grid search to do the cross-vaildation for the parameters tuning. for each top feature set, there are 84 sets for hyper-parameters combinations. You can find the detail of how I design the grid serach at SVM code cell<br>
 #feature 10
{'C': 100000, 'gamma': 0.0001, 'kernel': 'rbf'}
10 features  Spam precision: 0.9512195121951219
10 features  Spam recall 0.9312714776632303

 #feature100
{'C': 0.1, 'kernel': 'linear'}<br>
100 features  Spam precision: 0.9696969696969697<br>
100 features  Spam recall 0.9140893470790378<br>
 #feature1000
{'C': 0.1, 'kernel': 'linear'}<br>
1000 features  Spam precision: 0.8402777777777778<br>
1000 features  Spam recall 0.8419243986254296<br>
<br>
4.Adversarial Classification under NB classifier with binary<br>
Before attack:<br>
10 features  false negative rate BEFORE ad. modify: 0.04081632653061224<br>
10 features  Spam recall 0.9518900343642611<br>

After attack:<br>
average cost over all spam email:  1.7272727272727273<br>
accuracy recall after adversary strategy: 0.8419243986254296<br>
10 features  false negative rate AFTER ad. modify: 0.9387755102040817<br>

After classifer response:<br>
accuracy after classifier response 0.9037800687285223<br>
10 features  false negative rate AFTER ad. modify: 0.3673469387755102<>

